osu.Game.Rulesets.Osu/Skinning/Smoke.cs

```diff
@@ -0,0 +1,384 @@
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using osu.Framework.Allocation;
+using osu.Framework.Extensions.Color4Extensions;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Colour;
+using osu.Framework.Graphics.Primitives;
+using osu.Framework.Graphics.Rendering;
+using osu.Framework.Graphics.Rendering.Vertices;
+using osu.Framework.Graphics.Shaders;
+using osu.Framework.Graphics.Textures;
+using osu.Framework.Utils;
+using osu.Game.Rulesets.Osu.UI;
+using osuTK;
+using osuTK.Graphics;
+
+namespace osu.Game.Rulesets.Osu.Skinning
+{
+    public abstract class Smoke : Drawable, ITexturedShaderDrawable
+    {
+        public IShader? TextureShader { get; private set; }
+        public IShader? RoundedTextureShader { get; private set; }
+
+        private float? radius;
+
+        protected float Radius
+        {
+            get => radius ?? Texture?.DisplayWidth * 0.165f ?? 3;
+            set => radius = value;
+        }
+
+        protected Texture? Texture { get; set; }
+
+        protected double SmokeStartTime { get; private set; } = double.MinValue;
+
+        protected double SmokeEndTime { get; private set; } = double.MaxValue;
+
+        private Vector2 topLeft;
+
+        protected Vector2 TopLeft
+        {
+            get => topLeft;
+            set
+            {
+                if (topLeft == value)
+                    return;
+
+                topLeft = value;
+                Invalidate(Invalidation.Layout);
+            }
+        }
+
+        private Vector2 bottomRight;
+
+        protected Vector2 BottomRight
+        {
+            get => bottomRight;
+            set
+            {
+                if (bottomRight == value)
+                    return;
+
+                bottomRight = value;
+                Invalidate(Invalidation.Layout);
+            }
+        }
+
+        protected abstract double LifetimeAfterSmokeEnd { get; }
+        protected virtual float PointInterval => Radius * 7f / 8;
+        protected bool IsActive { get; private set; }
+
+        protected readonly List<SmokePoint> SmokePoints = new List<SmokePoint>();
+
+        private float totalDistance;
+        private Vector2? lastPosition;
+
+        private const int max_point_count = 18_000;
+
+        public override float Height
+        {
+            get => base.Height = BottomRight.Y - TopLeft.Y;
+            set => throw new InvalidOperationException($"Cannot manually set {nameof(Height)} of {nameof(Smoke)}.");
+        }
+
+        public override float Width
+        {
+            get => base.Width = BottomRight.X - TopLeft.X;
+            set => throw new InvalidOperationException($"Cannot manually set {nameof(Width)} of {nameof(Smoke)}.");
+        }
+
+        public override Vector2 Size
+        {
+            get => base.Size = BottomRight - TopLeft;
+            set => throw new InvalidOperationException($"Cannot manually set {nameof(Size)} of {nameof(Smoke)}.");
+        }
```

Comment:
Am I understanding right that this is supposed to make the smoke drawable only take up the amount of space of the smoke itself?

If so, I find this unnecessary - make the smoke take up the entire screen. Removes quite a bit of code here.

Reply:
Yep, you're correct. I did this because a) paths/sliders do this as well, and b) it's a bit annoying to use the visual inspector (ctrl f1) when drawable bounds don't visually fit.